### PR TITLE
[fix]: [CDS-73981]: Put validation on tag expression values in filters

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
Validate if the tag expression value is supported. Currently, only two expressions are supported - <+service.tags> and <+pipeline.tags>. <+pipline.tags> is resolved on UI Side. So if tag is an expression, then it should be <+service.tags>. Otherwise it is invalid & thus we throw a proper exception.

Screenshot 2023-07-24 at 6 10 06 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50916)

### CODE REVIEW

**Comment by lovish1999** [comment]
trigger uts

**Comment by tchaurasiya** [comment]
@lovish1999 can we add unit test for this?

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentInfraFilterHelper.java` line 695 at r1](https://reviewable.io/reviews/harness/harness-core/50916#-NaQ3LxY-TRFDRzqnmRW:-NaQ3LxY-TRFDRzqnmRX:b-p4i0wh) ([raw file](https://github.com/harness/harness-core/blob/166e4f7a23dcd348f92949f021d8419cb8053453/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentInfraFilterHelper.java#L695)):*
> ```Java
>     ParameterField<List<EnvironmentYamlV2>> environments = environmentGroup.getEnvironments();
>     if (ParameterField.isNotNull(environments)) {
>       validateTagExpressionValue((List<EnvironmentYamlV2>) environments);
> ```

environments in `ParameterField<List<EnvironmentYamlV2>>`, why are we casting it to `List<EnvironmentYamlV2>` ?

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentInfraFilterHelper.java` line 700 at r1](https://reviewable.io/reviews/harness/harness-core/50916#-NaQ39Fz-9XZtblf2KC8:-NaQ39Fz-9XZtblf2KC9:bl7mdly) ([raw file](https://github.com/harness/harness-core/blob/166e4f7a23dcd348f92949f021d8419cb8053453/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentInfraFilterHelper.java#L700)):*
> ```Java
> 
>   private boolean isSupportedTagExpression(String tagExpressionValue) {
>     return tagExpressionValue.equals(SERVICE_TAGS);
> ```

constants should always be on Left side of comparison
SERVICE_TAGS.equals(tagExpressionValue)

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentInfraFilterHelper.java` line 700 at r1](https://reviewable.io/reviews/harness/harness-core/50916#-NaQ3j7w9Mrp6uDQuL4y:-NaQ3j7w9Mrp6uDQuL4z:beyic97) ([raw file](https://github.com/harness/harness-core/blob/166e4f7a23dcd348f92949f021d8419cb8053453/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/environment/helper/EnvironmentInfraFilterHelper.java#L700)):*
> ```Java
> 
>   private boolean isSupportedTagExpression(String tagExpressionValue) {
>     return tagExpressionValue.equals(SERVICE_TAGS);
> ```

what happens when tagExpressionValue is empty ? Is it even possible ? Just asking

**Comment by yogesh-chauhan** [comment]
Should we not do this check in DeploymentStageFilterCreatorV2 i.e during pipeline save ?

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50916)* status: Some of the checks are too old. 2 file(s) to review. 0 of 2 approvals obtained.

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java` line 92 at r2](https://reviewable.io/reviews/harness/harness-core/50916#-NafUPbr-kbFL5KXx1FR:-NafUPbr-kbFL5KXx1FS:bm37b1d) ([raw file](https://github.com/harness/harness-core/blob/26bcecda735a1ddce7d7b9bfa06348d1129d89ee/125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java#L92)):*
> ```Java
>   private static final String STEP_IDENTIFIER_FIELD = "identifier";
>   private static final Set<String> SUPPORTED_FILTER_TAGS_EXPRESSIONS =
>       ImmutableSet.of("<+service.tags>", "<+pipeline.tags>");
> ```

nit: you can just use 
```
java.util.Set#of()
```

**Comment by lovish1999** [comment]
trigger uts

File: 125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java [hunk 1]
```diff
@@ -629,4 +637,43 @@ private List<String> findDuplicates(List<String> items) {
...
+  }
+
+  private void validateTagExpressionValue(ParameterField<List<FilterYaml>> filters) {
+    if (ParameterField.isNull(filters) || isEmpty(filters.getValue())) {
+      return;
+    }
+    for (FilterYaml filterYaml : filters.getValue()) {
+      if (FilterType.tags == filterYaml.getType()) {
+        ParameterField<Map<String, String>> tags = ((TagsFilter) filterYaml.getSpec()).getTags();
+        if (tags.isExpression() && !isSupportedTagExpression(tags.getExpressionValue())) {
```
Comment (tchaurasiya) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/creator/filters/DeploymentStageFilterJsonCreatorV2.java:667]: Can we handle the case where tags are mixed expression? <+pipeline.name>-suffix

**Review by tchaurasiya — CHANGES_REQUESTED** [review]

**Pull request closed without merging**